### PR TITLE
Feature/feature user pet images

### PR DESCRIPTION
**Pull request by AlbertPG**

### CODE REVIEW

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/activities/views/PetComponentView.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrObtainAllPetImages.java  2
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrUpdateUserImage.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/utilities/GetPetImageRunnable.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/TrUpdateUserImage.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/TrObtainAllPetImages.java  1
         

Complexity decreasing per file
==============================
+ app/src/main/java/org/pesmypetcare/mypetcare/services/PetManagerAdapter.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088&bid=17365756)

File: app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+public class ImageManager {
+    public static final String PET_PROFILE_IMAGES_PATH = "petProfileImages";
+    public static final String USER_PROFILE_IMAGES_PATH = "userProfileImages";
+    private static final String PATH = Environment.getExternalStorageDirectory().getAbsolutePath()
+        + File.separator + "MyPetCare";
+    private static final String EXTENSION = ".jpg";
+    private static final int QUALITY = 100;
+    private static Bitmap defaultPetImage;
+    private static byte[] defaultBytesPetImage;
```
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java:24]: Codacy Issue found: [Avoid unused private fields such as 'defaultBytesPetImage'.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

File: app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+    /**
+     * Read an image form a file.
+     * @param relativePath The path inside the folder with application data
+     * @param imageName The name of the image that has to be read
+     * @return The bytes of the image that has been read
+     * @throws IOException The file does not exist
+     */
+    public static byte[] readImage(String relativePath, String imageName) throws IOException {
+        FileInputStream input = new FileInputStream(PATH + File.separator + relativePath + File.separator
```
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java:76]: Codacy Issue found: [Ensure that resources like this FileInputStream object are closed after use](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

File: app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+     * Write an image to a file.
+     * @param relativePath The path inside the folder with application data
+     * @param fileName The name of the file that has to be written
+     * @param bytes The bytes that has to be written
+     */
+    public static void writeImage(String relativePath, String fileName, byte[] bytes) {
+        File file = new File(PATH + File.separator + relativePath);
+        file.mkdirs();
+        try {
+            FileOutputStream output = new FileOutputStream(PATH + File.separator + relativePath
```
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java:59]: Codacy Issue found: [Ensure that resources like this FileOutputStream object are closed after use](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

File: app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrObtainAllPetImages.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
+package org.pesmypetcare.mypetcare.controllers;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.pesmypetcare.mypetcare.features.pets.Pet;
+import org.pesmypetcare.mypetcare.features.pets.PetRepeatException;
```
Comment (santidrj) [inline app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrObtainAllPetImages.java:6]: Codacy Issue found: [Unused import - org.pesmypetcare.mypetcare.features.pets.PetRepeatException.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

File: app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrObtainAllPetImages.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    private User user;
+    private TrObtainAllPetImages trObtainAllPetImages;
+
+    @Before
+    public void setUp() {
+        user = new User("johnDoe2", "johndoe@gmail.com", "1234");
+        trObtainAllPetImages = new TrObtainAllPetImages(new StubPetManagerService());
+
+        for (int index = 0; index < 2; ++index) {
+            Pet pet = new Pet("pet" + index);
```
Comment (santidrj) [inline app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrObtainAllPetImages.java:25]: Codacy Issue found: [Avoid instantiating new objects inside loops](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

File: app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java (hunk 3 above)
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java:59]: Codacy Issue found: [Avoid instantiating FileInputStream, FileOutputStream, FileReader, or FileWriter](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

File: app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java (hunk 2 above)
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java:76]: Codacy Issue found: [Avoid instantiating FileInputStream, FileOutputStream, FileReader, or FileWriter](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

File: app/src/main/java/org/pesmypetcare/mypetcare/services/UserManagerAdapter.java [hunk 6]
```diff
@@ -12,20 +17,53 @@ public User findUserByUsername(String username) {
...
+     * Assign the user image from the server.
+     * @param user The user that has to be assigned an image
+     */
+    private void assignImageFromServer(User user) {
+        try {
+            byte[] userProfileImageBytes = ServiceLocator.getInstance().getUserManagerClient()
+                .downloadProfileImage(user.getToken(), user.getUsername());
+            user.setUserProfileImage(BitmapFactory.decodeByteArray(userProfileImageBytes, 0,
+                userProfileImageBytes.length));
+        } catch (ExecutionException | InterruptedException ignored) {
```
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/services/UserManagerAdapter.java:56]: Codacy Issue found: [Empty catch block.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- app/src/main/java/org/pesmypetcare/mypetcare/utilities/GetPetImageRunnable.java  2
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrObtainAllPetImages.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/TrUpdateUserImage.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/TrObtainAllPetImages.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/activities/views/PetComponentView.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrUpdateUserImage.java  1
         

Complexity decreasing per file
==============================
+ app/src/main/java/org/pesmypetcare/mypetcare/services/PetManagerAdapter.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088&bid=17365756)

File: app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java (hunk 3 above)
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java:59]: Codacy Issue found: [Ensure that resources like this FileOutputStream object are closed after use](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

File: app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java (hunk 2 above)
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java:76]: Codacy Issue found: [Avoid instantiating FileInputStream, FileOutputStream, FileReader, or FileWriter](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

File: app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrObtainAllPetImages.java (hunk 5 above)
Comment (santidrj) [inline app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrObtainAllPetImages.java:25]: Codacy Issue found: [Avoid instantiating new objects inside loops](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

File: app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java (hunk 2 above)
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java:76]: Codacy Issue found: [Ensure that resources like this FileInputStream object are closed after use](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

File: app/src/main/java/org/pesmypetcare/mypetcare/services/UserManagerAdapter.java (hunk 6 above)
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/services/UserManagerAdapter.java:56]: Codacy Issue found: [Empty catch block.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

File: app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java (hunk 1 above)
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java:24]: Codacy Issue found: [Avoid unused private fields such as 'defaultBytesPetImage'.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

File: app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrObtainAllPetImages.java (hunk 4 above)
Comment (santidrj) [inline app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrObtainAllPetImages.java:6]: Codacy Issue found: [Unused import - org.pesmypetcare.mypetcare.features.pets.PetRepeatException.](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

File: app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java (hunk 3 above)
Comment (santidrj) [inline app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java:59]: Codacy Issue found: [Avoid instantiating FileInputStream, FileOutputStream, FileReader, or FileWriter](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrUpdateUserImage.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/utilities/GetPetImageRunnable.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/TrObtainAllPetImages.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/TrUpdateUserImage.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrObtainAllPetImages.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/activities/views/PetComponentView.java  1
         

Complexity decreasing per file
==============================
+ app/src/main/java/org/pesmypetcare/mypetcare/services/PetManagerAdapter.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088&bid=17365756)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/main/java/org/pesmypetcare/mypetcare/activities/views/PetComponentView.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/TrUpdateUserImage.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrUpdateUserImage.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrObtainAllPetImages.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/utilities/GetPetImageRunnable.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/TrObtainAllPetImages.java  1
         

Complexity decreasing per file
==============================
+ app/src/main/java/org/pesmypetcare/mypetcare/services/PetManagerAdapter.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088&bid=17365756)

**Review by enrichp — APPROVED** [review]

**Review by xavier-campos-diaz — APPROVED** [review]

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/main/java/org/pesmypetcare/mypetcare/activities/views/PetComponentView.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/TrObtainAllPetImages.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/TrUpdateUserImage.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrObtainAllPetImages.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java  2
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrUpdateUserImage.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/utilities/GetPetImageRunnable.java  2
         

Complexity decreasing per file
==============================
+ app/src/main/java/org/pesmypetcare/mypetcare/services/PetManagerAdapter.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088&bid=17365756)

**Comment by santidrj** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/src/main/java/org/pesmypetcare/mypetcare/utilities/GetPetImageRunnable.java  2
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/TrObtainAllPetImages.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/activities/views/PetComponentView.java  1
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrObtainAllPetImages.java  2
- app/src/test/java/org/pesmypetcare/mypetcare/controllers/TestTrUpdateUserImage.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/controllers/TrUpdateUserImage.java  1
- app/src/main/java/org/pesmypetcare/mypetcare/utilities/ImageManager.java  2
         

Complexity decreasing per file
==============================
+ app/src/main/java/org/pesmypetcare/mypetcare/services/PetManagerAdapter.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Grupo13-PES-Mascotas/PES_My-Pet-Care/pullRequest?prid=5306088&bid=17365756)

**Pull request merged**